lib/srv/alpnproxy/proxy.go

```diff
@@ -0,0 +1,352 @@
+		c.ReadDeadline = defaults.HandshakeReadDeadline
+	}
+	return nil
+}
+
+// New creates a new instance of the Proxy.
+func New(cfg ProxyConfig) (*Proxy, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	var supportedProtocols []string
+	for k := range cfg.Router.alpnHandlers {
+		supportedProtocols = append(supportedProtocols, k)
+	}
+	return &Proxy{
+		cfg:                cfg,
+		log:                cfg.Log,
+		supportedProtocols: supportedProtocols,
+	}, nil
+}
+
+// Serve starts accepting connections.
+func (p *Proxy) Serve(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	p.cancel = cancel
+	p.cfg.TLSConfig.NextProtos = p.supportedProtocols
+	for {
+		clientConn, err := p.cfg.Listener.Accept()
+		if err != nil {
+			if utils.IsOKNetworkError(err) || trace.IsConnectionProblem(err) {
+				return nil
+			}
+			return trace.Wrap(err)
+		}
+		go func() {
+			if err := p.handleConn(ctx, clientConn); err != nil {
+				if err := clientConn.Close(); err != nil {
+					p.log.Warnf("failed to close client connection: %v", err)
+				}
+				p.log.Warnf("failed to handle client connection: %v", err)
+			}
+		}()
+	}
+}
+
+func (p *Proxy) handleConn(ctx context.Context, clientConn net.Conn) error {
+	hello, conn, err := p.readHelloMessageWithoutTLSTermination(clientConn)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	handlerDesc, err := p.getHandlerDescBaseOnClientHelloMsg(hello)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if handlerDesc.ForwardTLS {
+		return trace.Wrap(handlerDesc.Handler(ctx, conn))
+	}
+
+	tlsConn := tls.Server(conn, p.cfg.TLSConfig)
+	if err := tlsConn.SetReadDeadline(p.cfg.Clock.Now().Add(p.cfg.ReadDeadline)); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := tlsConn.Handshake(); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := tlsConn.SetReadDeadline(time.Time{}); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if p.isDatabaseConnection(tlsConn.ConnectionState()) {
+		return trace.Wrap(p.handleDatabaseConnection(ctx, tlsConn))
+	}
+	return trace.Wrap(handlerDesc.Handler(ctx, tlsConn))
+}
+
+// readHelloMessageWithoutTLSTermination allows reading a ClientHelloInfo message without termination of
+// incoming TLS connection. After calling readHelloMessageWithoutTLSTermination function a returned
+// net.Conn should be used for further operation.
+func (p *Proxy) readHelloMessageWithoutTLSTermination(conn net.Conn) (*tls.ClientHelloInfo, net.Conn, error) {
+	buff := new(bytes.Buffer)
+	var hello *tls.ClientHelloInfo
+	tlsConn := tls.Server(readOnlyConn{reader: io.TeeReader(conn, buff)}, &tls.Config{
+		GetConfigForClient: func(info *tls.ClientHelloInfo) (*tls.Config, error) {
+			hello = info
+			return nil, nil
+		},
+	})
+	if err := conn.SetReadDeadline(p.cfg.Clock.Now().Add(p.cfg.ReadDeadline)); err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+	err := tlsConn.Handshake()
+	if hello == nil {
+		return nil, nil, err
```

Comment:
trace.Wrap(err). Also, why is the handshake error returned only if hello is nil?

Reply:
I have added comment. 

This `err := tlsConn.Handshake()` call always returns an err because server certs are not provided but this allows to read a client helloMessage so this code is pre TLS handshake  performed only to get SNI ALPN value without TLS termination. The "real" handshake is preformed in handleConn or delegated to protocol handler if `ForwardTLS` flag was set in `HandlerDecs`